x/lightclient/ante/ibc_msg_update_client.go

```diff
@@ -1,82 +1,153 @@
 	"github.com/dymensionxyz/gerr-cosmos/gerrc"
 )
 
 func (i IBCMessagesDecorator) HandleMsgUpdateClient(ctx sdk.Context, msg *ibcclienttypes.MsgUpdateClient) error {
-	clientState, found := i.ibcClientKeeper.GetClientState(ctx, msg.ClientId)
-	if !found {
+	if !i.k.Enabled() {
 		return nil
 	}
-	// Cast client state to tendermint client state - we need this to get the chain id(rollapp id)
-	tmClientState, ok := clientState.(*ibctm.ClientState)
-	if !ok {
+	_, canonical := i.k.GetRollappForClientID(ctx, msg.ClientId)
+	header, err := getHeader(msg)
+	if !canonical && errorsmod.IsOf(err, errIsMisbehaviour) {
+		// We don't want to block misbehavior submission for non rollapps
+		return nil
+	}
+	if errorsmod.IsOf(err, errNoHeader) {
+		// it doesn't concern us
 		return nil
 	}
-	// Check if the client is the canonical client for the rollapp
-	rollappID := tmClientState.ChainId
-	canonicalClient, _ := i.lightClientKeeper.GetCanonicalClient(ctx, rollappID)
-	if canonicalClient != msg.ClientId {
-		return nil // The client is not a rollapp's canonical client. Continue with default behaviour.
+	if err != nil {
+		return errorsmod.Wrap(err, "get header")
+	}
+	seq, err := i.getSequencer(ctx, header)
+	err = errorsmod.Wrap(err, "get sequencer")
+	if errorsmod.IsOf(err, errProposerMismatch) {
+		// this should not occur on any chain, regardless of being a rollapp or not
+		return err
+	}
+	if errorsmod.IsOf(err, gerrc.ErrNotFound) {
+		if !canonical {
+			// not from sequencer, and not canonical - it's not interesting
+			return nil
+		}
+		return gerrc.ErrInvalidArgument.Wrap("update canonical client with non sequencer header")
+	}
+	if err != nil {
+		return err
+	}
+
+	// ~~~~~
+	// now we know that the msg is a header, and it was produced by a sequencer
+	// ~~~~~
+
+	if !seq.Bonded() {
+		// we assume here that sequencers will not propose blocks on other chains connected to the hub except for their rollapp
+		return gerrc.ErrInvalidArgument.Wrap("header is from unbonded sequencer")
+	}
+
+	rollapp, ok := i.raK.GetRollapp(ctx, seq.RollappId)
+	if !ok {
+		return gerrc.ErrInternal.Wrapf("get rollapp from sequencer: rollapp: %s", seq.RollappId)
+	}
+
+	// TODO: in hard fork will need to also use revision to make sure not from old revision
+	h := header.GetHeight().GetRevisionHeight()
+	stateInfos, err := i.getStateInfos(ctx, rollapp.RollappId, h)
+	if err != nil {
+		return errorsmod.Wrap(err, "get state infos")
+	}
+
+	if stateInfos.containingHPlus1 != nil {
+		// the header is pessimistic: the state update has already been received, so we check the header doesn't mismatch
+		return errorsmod.Wrap(i.validateUpdatePessimistically(ctx, stateInfos, header.ConsensusState(), h), "validate pessimistic")
+	}
+
+	// the header is optimistic: the state update has not yet been received, so we save optimistically
+	return errorsmod.Wrap(i.k.SaveSigner(ctx, seq.Address, msg.ClientId, h), "save updater")
```

Comment:
this wrapping of possible `nil` err is not intuitive
I would suggest to stick with convention (err check + wrap on error != nil)

Reply:
errorsmod is specifically designed to allow this

Reply:
> // If err is nil, this returns nil, avoiding the need for an if statement when
// wrapping a error returned at the end of a function